sdk/src/azure/core/az_hfsm_pipeline.c

```
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: MIT
+
+#include <azure/core/internal/az_hfsm_internal.h>
+#include <azure/core/internal/az_hfsm_pipeline_internal.h>
+#include <azure/core/az_platform.h>
+#include <azure/core/internal/az_precondition_internal.h>
+#include <azure/core/internal/az_result_internal.h>
+
+#include <azure/core/_az_cfg.h>
+
+AZ_NODISCARD az_result
+az_hfsm_pipeline_init(az_hfsm_pipeline* pipeline, az_hfsm_policy* outbound, az_hfsm_policy* inbound)
+{
+  pipeline->_internal.outbound_handler = outbound;
+  pipeline->_internal.inbound_handler = inbound;
+#ifndef TRANSPORT_MQTT_SYNC
+  return az_platform_mutex_init(&pipeline->_internal.mutex);
+#else
+  return AZ_OK;
+#endif
+}
+
+AZ_NODISCARD az_result
+az_hfsm_pipeline_post_outbound_event(az_hfsm_pipeline* pipeline, az_hfsm_event event)
+{
+  az_result ret;
+
+#ifndef TRANSPORT_MQTT_SYNC
+  _az_RETURN_IF_FAILED(az_platform_mutex_acquire(&pipeline->_internal.mutex));
+#endif
+
+  ret = az_hfsm_send_event((az_hfsm*)pipeline->_internal.outbound_handler, event);
+
+#ifndef TRANSPORT_MQTT_SYNC
+  _az_RETURN_IF_FAILED(az_platform_mutex_release(&pipeline->_internal.mutex));
+#endif
+
+  return ret;
+}
+
+AZ_NODISCARD az_result
+az_hfsm_pipeline_post_inbound_event(az_hfsm_pipeline* pipeline, az_hfsm_event event)
+{
+  az_result ret;
+
+#ifndef TRANSPORT_MQTT_SYNC
+  _az_RETURN_IF_FAILED(az_platform_mutex_acquire(&pipeline->_internal.mutex));
+#endif
+
+  ret = az_hfsm_send_event((az_hfsm*)pipeline->_internal.inbound_handler, event);
+
+#ifndef TRANSPORT_MQTT_SYNC
+  _az_RETURN_IF_FAILED(az_platform_mutex_release(&pipeline->_internal.mutex));
+#endif
+
+  return ret;
+}
+
+AZ_NODISCARD az_result
+az_hfsm_pipeline_send_inbound_event(az_hfsm_policy* policy, az_hfsm_event const event)
+{
+  az_result ret = az_hfsm_send_event((az_hfsm*)policy->inbound, event);
+  if (az_result_failed(ret))
+  {
+    // Replace the original event with an error event that is flowed to the application.
+    ret = az_hfsm_send_event(
```

just double checking here that the intended behavior is for the original error to be swallowed. If this succeeds, the return value for the function will be success, when the `send_event` failed.

---

This is correct, it is still possible for the function to return a failure depending on how the inbound policy state handles `AZ_HFSM_EVENT_ERROR`. `ret` will be set to whatever is returned from the handler.

---

@danewalton - To add more context, the inbound processing is performed on an MQTT callback stack. Returning an error to the MQTT callback is not very useful from the standpoint of the application. Instead, the error is flowed to the application callback _instead_ of the original event.

This is meant more as a catch-all for sync errors to ensure that the state machine and application understands there was a problem in processing the original event. 
An example would be a failure of the codec (i.e. JSON parsing error for example).